tests/runner/test_task_runner.py

```diff
@@ -0,0 +1,377 @@
+    s = sectionize(p, False)
+    t = TaskRunner(p)
+    t._prepare()
+    # make that do nothing
+    mocker.patch.object(t, "determine_max_slices")
+    s[0].max_slices = dummy_dataset.data.shape[0] / 2  # we'll have 2 blocks
+
+    # make this function return the block, with data multiplied by 2
+    def mul_block_by_two(section, block: DataSet):
+        block.data *= 2
+        return block
+
+    block_mock = mocker.patch.object(
+        t, "_execute_section_block", side_effect=mul_block_by_two
+    )
+
+    t._execute_section(s[0], 1)
+
+    np.testing.assert_allclose(t.dataset.data, original_value * 2)
+    calls = [call(ANY, ANY), call(ANY, ANY)]
+    block_mock.assert_has_calls(calls)  # make sure we got called twice
+
+
+def test_execute_section_for_block(mocker: MockerFixture, dummy_dataset: DataSet):
+    loader = make_test_loader(mocker)
+    loader.load.return_value = dummy_dataset
+    method1 = make_test_method(mocker, method_name="m1")
+    method2 = make_test_method(mocker, method_name="m2")
+    p = Pipeline(loader=loader, methods=[method1, method2],
+                 save_results_set=[False, False])
+    s = sectionize(p, False)
+    t = TaskRunner(p)
+    t._prepare()
+    exec_method = mocker.patch.object(t, "_execute_method", return_value=dummy_dataset)
+    t._execute_section_block(s[0], dummy_dataset)
+
+    calls = [call(method1, ANY, ANY), call(method2, ANY, ANY)]
+    exec_method.assert_has_calls(calls)
+
+
+@pytest.mark.parametrize("filebased", [False, True])
+def test_does_reslice_when_needed(
+    mocker: MockerFixture, dummy_dataset: DataSet, filebased: bool
+):
+    loader = make_test_loader(mocker)
+    loader.load.return_value = dummy_dataset
+    method1 = make_test_method(mocker, method_name="m1", pattern=Pattern.projection)
+    method2 = make_test_method(mocker, method_name="m2", pattern=Pattern.sinogram)
+    p = Pipeline(loader=loader, 
+                 methods=[method1, method2],
+                 save_results_set=[False, False])
+    reslice_dir = "./current" if filebased else None
+    t = TaskRunner(p, reslice_dir=reslice_dir)
+
+    exec_section = mocker.patch.object(t, "_execute_section")
+    reslice = mocker.patch("httomo.runner.task_runner.reslice")
+    reslice_filebased = mocker.patch("httomo.runner.task_runner.reslice_filebased")    
+
+    t.execute()
+
+    exec_section.assert_has_calls([call(ANY, 0), call(ANY, 1)])
+    assert t.reslice_count == 1
+    if filebased:
+        reslice_filebased.assert_called_once_with(
+                                                  ANY,
+                                                  1, 
+                                                  2,
+                                                  ANY,
+                                                  10,
+                                                  10,
+                                                  t.comm,
+                                                  reslice_dir)
```

Comment:
I think `ANY` was used in the assertion for the `data` parameter passed to `reslice_filebased()` rather than `dummy_dataset.data` because `dummy_dataset.data` actually gets set to the mock object during execution of the test due to:
- the patching of `reslice_filebased()`
- and the fact that in the task runner, the `data` attribute on the `DataSet` object is assigned to the return value of `reslice_filebased()` https://github.com/DiamondLightSource/httomo/pull/190/files#diff-d0196284273478c424c5d41d96d3d67518f3a6e9cda84e282da786d380407fe9R122
- hence, trying to assert that `dummy_dataset.data` was passed to `reslice_filebased()` doesn't work, because it's not a numpy array, but the mocked object instead

But the value for the `angles` parameter being passed to `reslice_filebased()` can still be asserted to be `dummy_dataset.angles` without issue since that doesn't get reset at any point during the execution of the test.

It makes sense to assert what we can, so asserting the angles should be done if it's simple enough, could you check if the `test_does_reslice_when_needed` test passes when using `dummy_dataset.angles` instead of `ANY`?

Also, some weird whitespace stuff going on here :stuck_out_tongue:

```suggestion
        reslice_filebased.assert_called_once_with(
            ANY,
            1, 
            2,
            dummy_dataset.angles,
            10,
            10,
            t.comm,
            reslice_dir
        )
```